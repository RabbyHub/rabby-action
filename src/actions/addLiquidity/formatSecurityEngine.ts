import {
  AddLiquidityRequireData,
  FormatSecurityEngineContext,
} from '../../types';

export const formatSecurityEngineAddLiquidity: FormatSecurityEngineContext =
  async (options) => {
    const { actionData, requireData, chainId } = options;

    if (!actionData.addLiquidity || !chainId) {
      return {} as any;
    }

    const data = requireData as AddLiquidityRequireData;

    return {
      addLiquidity: {
        receiverMatch:
          data.receiver?.toLowerCase() === data.sender.toLowerCase(),
        diff: data.diff,
      },
    };
  };
