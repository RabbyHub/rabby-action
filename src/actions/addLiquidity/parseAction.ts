import { ParseAction } from '../../types';
import { TokenItem } from '@rabby-wallet/rabby-api/dist/types';

interface AddLiquidityActionData {
  receiver: string;
  exchange_rate: string;
  min_exchange_rate: string;
  max_exchange_rate: string;
  token0: TokenItem;
  token1: TokenItem;
}

export const parseActionAddLiquidity: ParseAction = (options) => {
  const { data } = options;

  if (data?.type !== 'add_liquidity') {
    return {};
  }

  const {
    receiver,
    exchange_rate,
    min_exchange_rate,
    max_exchange_rate,
    token0,
    token1,
  } = data.data as unknown as AddLiquidityActionData;

  return {
    addLiquidity: {
      receiver,
      exchange_rate,
      min_exchange_rate,
      max_exchange_rate,
      token0,
      token1,
    },
  };
};
