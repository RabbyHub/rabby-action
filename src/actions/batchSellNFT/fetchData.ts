import { FetchActionRequiredData } from '../../types';
import { fetchDataAssetOrder } from '../assetOrder/fetchData';

export const fetchDataBatchSellNFT: FetchActionRequiredData = async (
  options
) => {
  if (options.type !== 'typed_data') {
    return {};
  }
  if (!options.actionData.contractId || !options.actionData.batchSellNFT) {
    return {};
  }

  return fetchDataAssetOrder(options, {
    receiver: options.actionData.contractId,
  });
};
