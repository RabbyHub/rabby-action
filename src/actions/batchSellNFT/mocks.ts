export const parseTxData = {
  action: {
    type: 'sell_nft_list_order',
    data: {
      pay_nft_list: [
        {
          id: 'c0645c0291405057c74674afda94ff2d',
          contract_id: '0x283c0bba69ebd4643cfce761b34b0206e75b2091',
          inner_id: '7293',
          chain: 'eth',
          symbol: 'SPRPUMA',
          name: '#7293',
          description:
            "Super PUMA is a collection of 10,000 PFPs representing variants from other timelines and universes that have made their way to the present after the events set in motion by Evil PUMA. \n\nSuper PUMA was originally a comic book originating in the 1970s. The character has been given a digital makeover that will see Super PUMA's story reinvigorated, highlighting historic moments in sport, but also pushing into new storytelling territory made possible by web3. Visit https://BlackStation.PUMA.com/SuperPUMA for more information. \n\nThe Super PUMA collection is complementary to PUMA's other ongoing NFT project the Nitro Collection, which will continue to be PUMA's vehicle to release new innovative shoe designs. \n\nNitro Collection: https://opensea.io/collection/nitrocollectionpuma \n\nTOS: https://blackstation.puma.com/terms-of-sale.pdf \n",
          content_type: 'image_url',
          content:
            'https://static.debank.com/image/eth_nft/local_url/c0645c0291405057c74674afda94ff2d/f4271ef8cdf77ce05dd69758fafa2232.png',
          thumbnail_url:
            'https://static.debank.com/image/eth_nft/thumbnail_url/c0645c0291405057c74674afda94ff2d/f4271ef8cdf77ce05dd69758fafa2232.png',
          total_supply: 1,
          attributes: [
            {
              trait_type: 'Background',
              value: 'Yellow',
              display_type: null,
            },
            {
              trait_type: 'Fur',
              value: 'Blue',
              display_type: null,
            },
            {
              trait_type: 'Item',
              value: 'Drumsticks',
              display_type: null,
            },
            {
              trait_type: 'Outfit',
              value: 'None',
              display_type: null,
            },
            {
              trait_type: 'Shoes',
              value: 'Red Suedes',
              display_type: null,
            },
            {
              trait_type: 'Head',
              value: 'Earbuds',
              display_type: null,
            },
            {
              trait_type: 'Face',
              value: 'Cool',
              display_type: null,
            },
          ],
          detail_url:
            'https://opensea.io/assets/0x283c0bba69ebd4643cfce761b34b0206e75b2091/7293',
          collection_id: 'eth:0x283c0bba69ebd4643cfce761b34b0206e75b2091',
          is_erc1155: false,
          is_erc721: true,
          pay_token: null,
          collection: {
            id: 'eth:0x283c0bba69ebd4643cfce761b34b0206e75b2091',
            chain: 'eth',
            name: 'Super PUMA',
            description: null,
            logo_url:
              'https://static.debank.com/image/nft_collection/logo_url/super-puma/7ce746a868adec939a7d5fd424825655.png',
            is_verified: true,
            credit_score: 790.9852183013668,
            is_scam: false,
            is_suspicious: false,
            is_core: true,
            floor_price: 0.0079,
          },
          amount: 1,
        },
        {
          id: '89ab90f922c61c933c76dd4887606c07',
          contract_id: '0x283c0bba69ebd4643cfce761b34b0206e75b2091',
          inner_id: '4884',
          chain: 'eth',
          symbol: 'SPRPUMA',
          name: '#4884',
          description:
            "Super PUMA is a collection of 10,000 PFPs representing variants from other timelines and universes that have made their way to the present after the events set in motion by Evil PUMA. \n\nSuper PUMA was originally a comic book originating in the 1970s. The character has been given a digital makeover that will see Super PUMA's story reinvigorated, highlighting historic moments in sport, but also pushing into new storytelling territory made possible by web3. Visit https://BlackStation.PUMA.com/SuperPUMA for more information. \n\nThe Super PUMA collection is complementary to PUMA's other ongoing NFT project the Nitro Collection, which will continue to be PUMA's vehicle to release new innovative shoe designs. \n\nNitro Collection: https://opensea.io/collection/nitrocollectionpuma \n\nTOS: https://blackstation.puma.com/terms-of-sale.pdf \n",
          content_type: 'image_url',
          content:
            'https://static.debank.com/image/eth_nft/local_url/89ab90f922c61c933c76dd4887606c07/9d5f9071e99df1e8b391457a875ba1aa.png',
          thumbnail_url:
            'https://static.debank.com/image/eth_nft/thumbnail_url/89ab90f922c61c933c76dd4887606c07/9d5f9071e99df1e8b391457a875ba1aa.png',
          total_supply: 1,
          attributes: [
            {
              trait_type: 'Background',
              value: 'Pink',
              display_type: null,
            },
            {
              trait_type: 'Fur',
              value: 'Tie-dye',
              display_type: null,
            },
            {
              trait_type: 'Item',
              value: 'Cobra driver',
              display_type: null,
            },
            {
              trait_type: 'Outfit',
              value: 'Shorts',
              display_type: null,
            },
            {
              trait_type: 'Shoes',
              value: 'NFRNO',
              display_type: null,
            },
            {
              trait_type: 'Head',
              value: 'Earbuds',
              display_type: null,
            },
            {
              trait_type: 'Face',
              value: 'Cool',
              display_type: null,
            },
          ],
          detail_url:
            'https://opensea.io/assets/0x283c0bba69ebd4643cfce761b34b0206e75b2091/4884',
          collection_id: 'eth:0x283c0bba69ebd4643cfce761b34b0206e75b2091',
          is_erc1155: false,
          is_erc721: true,
          pay_token: null,
          collection: {
            id: 'eth:0x283c0bba69ebd4643cfce761b34b0206e75b2091',
            chain: 'eth',
            name: 'Super PUMA',
            description: null,
            logo_url:
              'https://static.debank.com/image/nft_collection/logo_url/super-puma/7ce746a868adec939a7d5fd424825655.png',
            is_verified: true,
            credit_score: 790.9852183013668,
            is_scam: false,
            is_suspicious: false,
            is_core: true,
            floor_price: 0.0079,
          },
          amount: 1,
        },
      ],
      receive_token: {
        id: '0xfe1ef2b469846d1832b25095ff51b004f090e0c6',
        chain: 'eth',
        name: 'Pepe Coin',
        symbol: 'PEPE',
        display_symbol: null,
        optimized_symbol: 'PEPE',
        decimals: 18,
        logo_url:
          'https://static.debank.com/image/eth_token/logo_url/0xfe1ef2b469846d1832b25095ff51b004f090e0c6/6d16ba4370ad7b736d32338204a4fef3.png',
        protocol_id: '',
        price: 3.5794827932828884e-10,
        price_24h_change: null,
        credit_score: 4268.682697297744,
        is_verified: false,
        is_scam: true,
        is_suspicious: false,
        is_core: false,
        is_wallet: false,
        time_at: 1684097147.0,
        amount: 0.0,
        raw_amount: '0',
      },
      receiver: '0xf08c90c7f470b640a21dd9b3744eca3d1d16a044',
      takers: [],
      expire_at: 1690962184,
    },
    is_asset_changed: true,
    is_involving_privacy: false,
  },
  log_id: 6318841,
} as any;

export const txData = {
  types: {
    EIP712Domain: [
      {
        name: 'name',
        type: 'string',
      },
      {
        name: 'version',
        type: 'string',
      },
      {
        name: 'chainId',
        type: 'uint256',
      },
      {
        name: 'verifyingContract',
        type: 'address',
      },
    ],
    BulkOrder: [
      {
        name: 'tree',
        type: 'OrderComponents[2]',
      },
    ],
    OrderComponents: [
      {
        name: 'offerer',
        type: 'address',
      },
      {
        name: 'zone',
        type: 'address',
      },
      {
        name: 'offer',
        type: 'OfferItem[]',
      },
      {
        name: 'consideration',
        type: 'ConsiderationItem[]',
      },
      {
        name: 'orderType',
        type: 'uint8',
      },
      {
        name: 'startTime',
        type: 'uint256',
      },
      {
        name: 'endTime',
        type: 'uint256',
      },
      {
        name: 'zoneHash',
        type: 'bytes32',
      },
      {
        name: 'salt',
        type: 'uint256',
      },
      {
        name: 'conduitKey',
        type: 'bytes32',
      },
      {
        name: 'counter',
        type: 'uint256',
      },
    ],
    OfferItem: [
      {
        name: 'itemType',
        type: 'uint8',
      },
      {
        name: 'token',
        type: 'address',
      },
      {
        name: 'identifierOrCriteria',
        type: 'uint256',
      },
      {
        name: 'startAmount',
        type: 'uint256',
      },
      {
        name: 'endAmount',
        type: 'uint256',
      },
    ],
    ConsiderationItem: [
      {
        name: 'itemType',
        type: 'uint8',
      },
      {
        name: 'token',
        type: 'address',
      },
      {
        name: 'identifierOrCriteria',
        type: 'uint256',
      },
      {
        name: 'startAmount',
        type: 'uint256',
      },
      {
        name: 'endAmount',
        type: 'uint256',
      },
      {
        name: 'recipient',
        type: 'address',
      },
    ],
  },
  primaryType: 'BulkOrder',
  domain: {
    name: 'Seaport',
    version: '1.5',
    chainId: '1',
    verifyingContract: '0x00000000000000adc04c56bf30ac9d3c0aaf14dc',
  },
  message: {
    tree: [
      {
        offerer: '0xF08C90C7f470B640a21DD9B3744eca3d1d16a044',
        offer: [
          {
            itemType: '2',
            token: '0x283c0BbA69EbD4643cfCe761B34b0206E75b2091',
            identifierOrCriteria: '7293',
            startAmount: '1',
            endAmount: '1',
          },
        ],
        consideration: [
          {
            itemType: '1',
            token: '0xfe1ef2b469846d1832b25095ff51b004f090e0c6',
            identifierOrCriteria: '0',
            startAmount: '925000000000000000',
            endAmount: '925000000000000000',
            recipient: '0x5853ed4f26a3fcea565b3fbc698bb19cdf6deb85',
          },
          {
            itemType: '1',
            token: '0xfe1ef2b469846d1832b25095ff51b004f090e0c6',
            identifierOrCriteria: '0',
            startAmount: '25000000000000000',
            endAmount: '25000000000000000',
            recipient: '0x0000a26b00c1F0DF003000390027140000fAa719',
          },
          {
            itemType: '1',
            token: '0xfe1ef2b469846d1832b25095ff51b004f090e0c6',
            identifierOrCriteria: '0',
            startAmount: '50000000000000000',
            endAmount: '50000000000000000',
            recipient: '0xC92DfAF17B5F79149668A638F137d7090Ca740Ae',
          },
        ],
        startTime: '1688283784',
        endTime: '1690962184',
        orderType: '0',
        zone: '0x004C00500000aD104D7DBd00e3ae0A5C00560C00',
        zoneHash:
          '0x0000000000000000000000000000000000000000000000000000000000000000',
        salt: '24446860302761739304752683030156737591518664810215442929805511063122729690650',
        conduitKey:
          '0x0000007b02230091a7ed01230072f7006a004d60a8d4e71d599b8104250f0000',
        totalOriginalConsiderationItems: '3',
        counter: '0',
      },
      {
        offerer: '0xF08C90C7f470B640a21DD9B3744eca3d1d16a044',
        offer: [
          {
            itemType: '2',
            token: '0x283c0BbA69EbD4643cfCe761B34b0206E75b2091',
            identifierOrCriteria: '4884',
            startAmount: '1',
            endAmount: '1',
          },
        ],
        consideration: [
          {
            itemType: '1',
            token: '0xfe1ef2b469846d1832b25095ff51b004f090e0c6',
            identifierOrCriteria: '0',
            startAmount: '925000000000000000',
            endAmount: '925000000000000000',
            recipient: '0x5853ed4f26a3fcea565b3fbc698bb19cdf6deb85',
          },
          {
            itemType: '1',
            token: '0xfe1ef2b469846d1832b25095ff51b004f090e0c6',
            identifierOrCriteria: '0',
            startAmount: '25000000000000000',
            endAmount: '25000000000000000',
            recipient: '0x0000a26b00c1F0DF003000390027140000fAa719',
          },
          {
            itemType: '1',
            token: '0xfe1ef2b469846d1832b25095ff51b004f090e0c6',
            identifierOrCriteria: '0',
            startAmount: '50000000000000000',
            endAmount: '50000000000000000',
            recipient: '0xC92DfAF17B5F79149668A638F137d7090Ca740Ae',
          },
        ],
        startTime: '1688283784',
        endTime: '1690962184',
        orderType: '0',
        zone: '0x004C00500000aD104D7DBd00e3ae0A5C00560C00',
        zoneHash:
          '0x0000000000000000000000000000000000000000000000000000000000000000',
        salt: '24446860302761739304752683030156737591518664810215442929806425803700743317776',
        conduitKey:
          '0x0000007b02230091a7ed01230072f7006a004d60a8d4e71d599b8104250f0000',
        totalOriginalConsiderationItems: '3',
        counter: '0',
      },
    ],
  },
};
