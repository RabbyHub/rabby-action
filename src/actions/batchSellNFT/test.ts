import {
  apiProvider,
  formatProvider,
  walletProvider,
  ETH_CHAIN_ID,
  ORIGIN,
} from '../../../__mocks__';
import { parseTxData, txData } from './mocks';
import { parseActionBatchSellNFT } from './parseAction';
import { fetchDataBatchSellNFT } from './fetchData';
import { formatSecurityEngineBatchSellNFT } from './formatSecurityEngine';
import { parseTypedDataAction } from '../../utils/parseTypedDataAction';

/**
 * https://brain.debank.com/action-group/typed_data/32ae1e7b9e9e249e2faea61155fb20a0
 * userId: 0xcb1605ed17f6145db16e26e1d8adfe8f4b175377
 * chainId: eth
 */
test('BatchSellNFT', async () => {
  const actionData = parseTypedDataAction(parseActionBatchSellNFT)({
    type: 'typed_data',
    data: parseTxData['action'],
    typedData: txData,
    sender: '0xcb1605ed17f6145db16e26e1d8adfe8f4b175377',
  });
  expect(actionData).toMatchSnapshot('parseActionBatchSellNFT');

  const requireData = await fetchDataBatchSellNFT({
    type: 'typed_data',
    actionData,
    chainId: ETH_CHAIN_ID,
    walletProvider,
    apiProvider,
    sender: '0xcb1605ed17f6145db16e26e1d8adfe8f4b175377',
  });
  expect(requireData).toMatchSnapshot('fetchDataBatchSellNFT');

  const ctx = await formatSecurityEngineBatchSellNFT({
    type: 'typed_data',
    actionData,
    requireData,
    chainId: ETH_CHAIN_ID,
    isTestnet: false,
    provider: formatProvider,
    origin: ORIGIN,
  });
  expect(ctx).toMatchSnapshot('formatSecurityEngineBatchSellNFT');
});
