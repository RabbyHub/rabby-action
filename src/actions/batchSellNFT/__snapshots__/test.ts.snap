// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BatchSellNFT: fetchDataBatchSellNFT 1`] = `
{
  "bornAt": 1591388241,
  "extraState": {
    "hasInteraction": [Function],
    "receiverHasTransfer": [Function],
  },
  "hasInteraction": true,
  "id": "0x00000000000000adc04c56bf30ac9d3c0aaf14dc",
  "protocol": {
    "id": "uniswap2",
    "logo_url": "https://static.debank.com/image/project/logo_url/uniswap2/87a541b3b83b041c8d12119e5a0d19f0.png",
    "name": "Uniswap V2",
  },
  "rank": 65,
  "receiverInWallet": false,
  "sender": "0xcb1605ed17f6145db16e26e1d8adfe8f4b175377",
  "unexpectedAddr": null,
}
`;

exports[`BatchSellNFT: fetchDataBatchSellNFT 2`] = `
{
  "bornAt": 1591388241,
  "extraState": {
    "hasInteraction": [Function],
    "receiverHasTransfer": [Function],
  },
  "hasInteraction": true,
  "id": "0x00000000000000adc04c56bf30ac9d3c0aaf14dc",
  "protocol": {
    "id": "uniswap2",
    "logo_url": "https://static.debank.com/image/project/logo_url/uniswap2/87a541b3b83b041c8d12119e5a0d19f0.png",
    "name": "Uniswap V2",
  },
  "rank": 65,
  "receiverInWallet": false,
  "sender": "0xcb1605ed17f6145db16e26e1d8adfe8f4b175377",
  "unexpectedAddr": null,
}
`;

exports[`BatchSellNFT: formatSecurityEngineBatchSellNFT 1`] = `
{
  "batchSellNFT": {
    "chainId": "eth",
    "from": "0xcb1605ed17f6145db16e26e1d8adfe8f4b175377",
    "id": "0x00000000000000adc04c56bf30ac9d3c0aaf14dc",
    "receiveTokenHasFake": true,
    "receiveTokenHasScam": false,
    "receiver": "0xf08c90c7f470b640a21dd9b3744eca3d1d16a044",
    "specificBuyer": undefined,
  },
}
`;

exports[`BatchSellNFT: formatSecurityEngineBatchSellNFT 2`] = `
{
  "batchSellNFT": {
    "chainId": "eth",
    "from": "0xcb1605ed17f6145db16e26e1d8adfe8f4b175377",
    "id": "0x00000000000000adc04c56bf30ac9d3c0aaf14dc",
    "receiveTokenHasFake": true,
    "receiveTokenHasScam": false,
    "receiver": "0xf08c90c7f470b640a21dd9b3744eca3d1d16a044",
    "specificBuyer": undefined,
  },
}
`;

exports[`BatchSellNFT: parseActionBatchSellNFT 1`] = `
{
  "actionType": "sell_nft_list_order",
  "batchSellNFT": {
    "expire_at": 1690962184,
    "pay_nft_list": [
      {
        "amount": 1,
        "attributes": [
          {
            "display_type": null,
            "trait_type": "Background",
            "value": "Yellow",
          },
          {
            "display_type": null,
            "trait_type": "Fur",
            "value": "Blue",
          },
          {
            "display_type": null,
            "trait_type": "Item",
            "value": "Drumsticks",
          },
          {
            "display_type": null,
            "trait_type": "Outfit",
            "value": "None",
          },
          {
            "display_type": null,
            "trait_type": "Shoes",
            "value": "Red Suedes",
          },
          {
            "display_type": null,
            "trait_type": "Head",
            "value": "Earbuds",
          },
          {
            "display_type": null,
            "trait_type": "Face",
            "value": "Cool",
          },
        ],
        "chain": "eth",
        "collection": {
          "chain": "eth",
          "credit_score": 790.9852183013668,
          "description": null,
          "floor_price": 0.0079,
          "id": "eth:0x283c0bba69ebd4643cfce761b34b0206e75b2091",
          "is_core": true,
          "is_scam": false,
          "is_suspicious": false,
          "is_verified": true,
          "logo_url": "https://static.debank.com/image/nft_collection/logo_url/super-puma/7ce746a868adec939a7d5fd424825655.png",
          "name": "Super PUMA",
        },
        "collection_id": "eth:0x283c0bba69ebd4643cfce761b34b0206e75b2091",
        "content": "https://static.debank.com/image/eth_nft/local_url/c0645c0291405057c74674afda94ff2d/f4271ef8cdf77ce05dd69758fafa2232.png",
        "content_type": "image_url",
        "contract_id": "0x283c0bba69ebd4643cfce761b34b0206e75b2091",
        "description": "Super PUMA is a collection of 10,000 PFPs representing variants from other timelines and universes that have made their way to the present after the events set in motion by Evil PUMA. 

Super PUMA was originally a comic book originating in the 1970s. The character has been given a digital makeover that will see Super PUMA's story reinvigorated, highlighting historic moments in sport, but also pushing into new storytelling territory made possible by web3. Visit https://BlackStation.PUMA.com/SuperPUMA for more information. 

The Super PUMA collection is complementary to PUMA's other ongoing NFT project the Nitro Collection, which will continue to be PUMA's vehicle to release new innovative shoe designs. 

Nitro Collection: https://opensea.io/collection/nitrocollectionpuma 

TOS: https://blackstation.puma.com/terms-of-sale.pdf 
",
        "detail_url": "https://opensea.io/assets/0x283c0bba69ebd4643cfce761b34b0206e75b2091/7293",
        "id": "c0645c0291405057c74674afda94ff2d",
        "inner_id": "7293",
        "is_erc1155": false,
        "is_erc721": true,
        "name": "#7293",
        "pay_token": null,
        "symbol": "SPRPUMA",
        "thumbnail_url": "https://static.debank.com/image/eth_nft/thumbnail_url/c0645c0291405057c74674afda94ff2d/f4271ef8cdf77ce05dd69758fafa2232.png",
        "total_supply": 1,
      },
      {
        "amount": 1,
        "attributes": [
          {
            "display_type": null,
            "trait_type": "Background",
            "value": "Pink",
          },
          {
            "display_type": null,
            "trait_type": "Fur",
            "value": "Tie-dye",
          },
          {
            "display_type": null,
            "trait_type": "Item",
            "value": "Cobra driver",
          },
          {
            "display_type": null,
            "trait_type": "Outfit",
            "value": "Shorts",
          },
          {
            "display_type": null,
            "trait_type": "Shoes",
            "value": "NFRNO",
          },
          {
            "display_type": null,
            "trait_type": "Head",
            "value": "Earbuds",
          },
          {
            "display_type": null,
            "trait_type": "Face",
            "value": "Cool",
          },
        ],
        "chain": "eth",
        "collection": {
          "chain": "eth",
          "credit_score": 790.9852183013668,
          "description": null,
          "floor_price": 0.0079,
          "id": "eth:0x283c0bba69ebd4643cfce761b34b0206e75b2091",
          "is_core": true,
          "is_scam": false,
          "is_suspicious": false,
          "is_verified": true,
          "logo_url": "https://static.debank.com/image/nft_collection/logo_url/super-puma/7ce746a868adec939a7d5fd424825655.png",
          "name": "Super PUMA",
        },
        "collection_id": "eth:0x283c0bba69ebd4643cfce761b34b0206e75b2091",
        "content": "https://static.debank.com/image/eth_nft/local_url/89ab90f922c61c933c76dd4887606c07/9d5f9071e99df1e8b391457a875ba1aa.png",
        "content_type": "image_url",
        "contract_id": "0x283c0bba69ebd4643cfce761b34b0206e75b2091",
        "description": "Super PUMA is a collection of 10,000 PFPs representing variants from other timelines and universes that have made their way to the present after the events set in motion by Evil PUMA. 

Super PUMA was originally a comic book originating in the 1970s. The character has been given a digital makeover that will see Super PUMA's story reinvigorated, highlighting historic moments in sport, but also pushing into new storytelling territory made possible by web3. Visit https://BlackStation.PUMA.com/SuperPUMA for more information. 

The Super PUMA collection is complementary to PUMA's other ongoing NFT project the Nitro Collection, which will continue to be PUMA's vehicle to release new innovative shoe designs. 

Nitro Collection: https://opensea.io/collection/nitrocollectionpuma 

TOS: https://blackstation.puma.com/terms-of-sale.pdf 
",
        "detail_url": "https://opensea.io/assets/0x283c0bba69ebd4643cfce761b34b0206e75b2091/4884",
        "id": "89ab90f922c61c933c76dd4887606c07",
        "inner_id": "4884",
        "is_erc1155": false,
        "is_erc721": true,
        "name": "#4884",
        "pay_token": null,
        "symbol": "SPRPUMA",
        "thumbnail_url": "https://static.debank.com/image/eth_nft/thumbnail_url/89ab90f922c61c933c76dd4887606c07/9d5f9071e99df1e8b391457a875ba1aa.png",
        "total_supply": 1,
      },
    ],
    "receive_token": {
      "amount": 0,
      "chain": "eth",
      "credit_score": 4268.682697297744,
      "decimals": 18,
      "display_symbol": null,
      "id": "0xfe1ef2b469846d1832b25095ff51b004f090e0c6",
      "is_core": false,
      "is_scam": true,
      "is_suspicious": false,
      "is_verified": false,
      "is_wallet": false,
      "logo_url": "https://static.debank.com/image/eth_token/logo_url/0xfe1ef2b469846d1832b25095ff51b004f090e0c6/6d16ba4370ad7b736d32338204a4fef3.png",
      "name": "Pepe Coin",
      "optimized_symbol": "PEPE",
      "price": 3.5794827932828884e-10,
      "price_24h_change": null,
      "protocol_id": "",
      "raw_amount": "0",
      "symbol": "PEPE",
      "time_at": 1684097147,
    },
    "receiver": "0xf08c90c7f470b640a21dd9b3744eca3d1d16a044",
    "takers": [],
  },
  "brand": undefined,
  "chainId": "1",
  "contractId": "0x00000000000000adc04c56bf30ac9d3c0aaf14dc",
  "sender": "0xcb1605ed17f6145db16e26e1d8adfe8f4b175377",
}
`;

exports[`BatchSellNFT: parseActionBatchSellNFT 2`] = `
{
  "actionType": "sell_nft_list_order",
  "batchSellNFT": {
    "expire_at": 1690962184,
    "pay_nft_list": [
      {
        "amount": 1,
        "attributes": [
          {
            "display_type": null,
            "trait_type": "Background",
            "value": "Yellow",
          },
          {
            "display_type": null,
            "trait_type": "Fur",
            "value": "Blue",
          },
          {
            "display_type": null,
            "trait_type": "Item",
            "value": "Drumsticks",
          },
          {
            "display_type": null,
            "trait_type": "Outfit",
            "value": "None",
          },
          {
            "display_type": null,
            "trait_type": "Shoes",
            "value": "Red Suedes",
          },
          {
            "display_type": null,
            "trait_type": "Head",
            "value": "Earbuds",
          },
          {
            "display_type": null,
            "trait_type": "Face",
            "value": "Cool",
          },
        ],
        "chain": "eth",
        "collection": {
          "chain": "eth",
          "credit_score": 790.9852183013668,
          "description": null,
          "floor_price": 0.0079,
          "id": "eth:0x283c0bba69ebd4643cfce761b34b0206e75b2091",
          "is_core": true,
          "is_scam": false,
          "is_suspicious": false,
          "is_verified": true,
          "logo_url": "https://static.debank.com/image/nft_collection/logo_url/super-puma/7ce746a868adec939a7d5fd424825655.png",
          "name": "Super PUMA",
        },
        "collection_id": "eth:0x283c0bba69ebd4643cfce761b34b0206e75b2091",
        "content": "https://static.debank.com/image/eth_nft/local_url/c0645c0291405057c74674afda94ff2d/f4271ef8cdf77ce05dd69758fafa2232.png",
        "content_type": "image_url",
        "contract_id": "0x283c0bba69ebd4643cfce761b34b0206e75b2091",
        "description": "Super PUMA is a collection of 10,000 PFPs representing variants from other timelines and universes that have made their way to the present after the events set in motion by Evil PUMA. 

Super PUMA was originally a comic book originating in the 1970s. The character has been given a digital makeover that will see Super PUMA's story reinvigorated, highlighting historic moments in sport, but also pushing into new storytelling territory made possible by web3. Visit https://BlackStation.PUMA.com/SuperPUMA for more information. 

The Super PUMA collection is complementary to PUMA's other ongoing NFT project the Nitro Collection, which will continue to be PUMA's vehicle to release new innovative shoe designs. 

Nitro Collection: https://opensea.io/collection/nitrocollectionpuma 

TOS: https://blackstation.puma.com/terms-of-sale.pdf 
",
        "detail_url": "https://opensea.io/assets/0x283c0bba69ebd4643cfce761b34b0206e75b2091/7293",
        "id": "c0645c0291405057c74674afda94ff2d",
        "inner_id": "7293",
        "is_erc1155": false,
        "is_erc721": true,
        "name": "#7293",
        "pay_token": null,
        "symbol": "SPRPUMA",
        "thumbnail_url": "https://static.debank.com/image/eth_nft/thumbnail_url/c0645c0291405057c74674afda94ff2d/f4271ef8cdf77ce05dd69758fafa2232.png",
        "total_supply": 1,
      },
      {
        "amount": 1,
        "attributes": [
          {
            "display_type": null,
            "trait_type": "Background",
            "value": "Pink",
          },
          {
            "display_type": null,
            "trait_type": "Fur",
            "value": "Tie-dye",
          },
          {
            "display_type": null,
            "trait_type": "Item",
            "value": "Cobra driver",
          },
          {
            "display_type": null,
            "trait_type": "Outfit",
            "value": "Shorts",
          },
          {
            "display_type": null,
            "trait_type": "Shoes",
            "value": "NFRNO",
          },
          {
            "display_type": null,
            "trait_type": "Head",
            "value": "Earbuds",
          },
          {
            "display_type": null,
            "trait_type": "Face",
            "value": "Cool",
          },
        ],
        "chain": "eth",
        "collection": {
          "chain": "eth",
          "credit_score": 790.9852183013668,
          "description": null,
          "floor_price": 0.0079,
          "id": "eth:0x283c0bba69ebd4643cfce761b34b0206e75b2091",
          "is_core": true,
          "is_scam": false,
          "is_suspicious": false,
          "is_verified": true,
          "logo_url": "https://static.debank.com/image/nft_collection/logo_url/super-puma/7ce746a868adec939a7d5fd424825655.png",
          "name": "Super PUMA",
        },
        "collection_id": "eth:0x283c0bba69ebd4643cfce761b34b0206e75b2091",
        "content": "https://static.debank.com/image/eth_nft/local_url/89ab90f922c61c933c76dd4887606c07/9d5f9071e99df1e8b391457a875ba1aa.png",
        "content_type": "image_url",
        "contract_id": "0x283c0bba69ebd4643cfce761b34b0206e75b2091",
        "description": "Super PUMA is a collection of 10,000 PFPs representing variants from other timelines and universes that have made their way to the present after the events set in motion by Evil PUMA. 

Super PUMA was originally a comic book originating in the 1970s. The character has been given a digital makeover that will see Super PUMA's story reinvigorated, highlighting historic moments in sport, but also pushing into new storytelling territory made possible by web3. Visit https://BlackStation.PUMA.com/SuperPUMA for more information. 

The Super PUMA collection is complementary to PUMA's other ongoing NFT project the Nitro Collection, which will continue to be PUMA's vehicle to release new innovative shoe designs. 

Nitro Collection: https://opensea.io/collection/nitrocollectionpuma 

TOS: https://blackstation.puma.com/terms-of-sale.pdf 
",
        "detail_url": "https://opensea.io/assets/0x283c0bba69ebd4643cfce761b34b0206e75b2091/4884",
        "id": "89ab90f922c61c933c76dd4887606c07",
        "inner_id": "4884",
        "is_erc1155": false,
        "is_erc721": true,
        "name": "#4884",
        "pay_token": null,
        "symbol": "SPRPUMA",
        "thumbnail_url": "https://static.debank.com/image/eth_nft/thumbnail_url/89ab90f922c61c933c76dd4887606c07/9d5f9071e99df1e8b391457a875ba1aa.png",
        "total_supply": 1,
      },
    ],
    "receive_token": {
      "amount": 0,
      "chain": "eth",
      "credit_score": 4268.682697297744,
      "decimals": 18,
      "display_symbol": null,
      "id": "0xfe1ef2b469846d1832b25095ff51b004f090e0c6",
      "is_core": false,
      "is_scam": true,
      "is_suspicious": false,
      "is_verified": false,
      "is_wallet": false,
      "logo_url": "https://static.debank.com/image/eth_token/logo_url/0xfe1ef2b469846d1832b25095ff51b004f090e0c6/6d16ba4370ad7b736d32338204a4fef3.png",
      "name": "Pepe Coin",
      "optimized_symbol": "PEPE",
      "price": 3.5794827932828884e-10,
      "price_24h_change": null,
      "protocol_id": "",
      "raw_amount": "0",
      "symbol": "PEPE",
      "time_at": 1684097147,
    },
    "receiver": "0xf08c90c7f470b640a21dd9b3744eca3d1d16a044",
    "takers": [],
  },
  "brand": undefined,
  "chainId": "1",
  "contractId": "0x00000000000000adc04c56bf30ac9d3c0aaf14dc",
  "sender": "0xcb1605ed17f6145db16e26e1d8adfe8f4b175377",
}
`;
